src/redprl/refiner.fun

```diff
@@ -708,32 +754,57 @@ struct
          | _ => raise E.error [Fpp.text "Could not find suitable type synthesis rule for", TermPrinter.ppTerm m]
 
       fun StepJdg sign = matchGoal
-        (fn _ >> CJ.TRUE ty => StepTrue sign ty
-          | _ >> CJ.EQ_TYPE tys => StepEqType sign tys
-          | _ >> CJ.EQ ((m, n), ty) => StepEq sign ((m, n), ty)
-          | _ >> CJ.SYNTH m => StepSynth sign m
+        (fn _ >> CJ.TRUE (ty, _) => StepTrue sign ty
+          | _ >> CJ.EQ_TYPE (tys, _) => StepEqType sign tys
+          | _ >> CJ.EQ ((m, n), (ty, _)) => StepEq sign ((m, n), ty)
+          | _ >> CJ.SYNTH (m, _) => StepSynth sign m
           | _ >> CJ.PARAM_SUBST _ => Misc.ParamSubst
           | MATCH _ => Misc.MatchOperator
           | MATCH_RECORD _ => Record.MatchRecord orelse_ Computation.MatchRecordHeadExpansion sign then_ Record.MatchRecord)
 
+      fun EqTypeFromHyp alpha jdg =
+        let
+          val (_, H) >> CJ.EQ_TYPE ((a0, b0), k0) = jdg
+          val isUnary = Abt.eq (a0, b0)
+          val isUseful =
+            fn CJ.EQ_TYPE ((a1, b1), k1) =>
+                 Abt.eq (a0, a1) andalso Abt.eq (b0, b1)
+                 andalso K.greatestMeetRight' (SOME k0, SOME k1) <> SOME k0
+             | CJ.EQ (_, (a1, k1)) =>
+                 isUnary andalso Abt.eq (a0, a1)
+                 andalso K.greatestMeetRight' (SOME k0, SOME k1) <> SOME k0
+             | CJ.TRUE (a1, k1) =>
+                 isUnary andalso Abt.eq (a0, a1)
+                 andalso K.greatestMeetRight' (SOME k0, SOME k1) <> SOME k0
+             | _ => false
+        in
+          case Hyps.search H isUseful of
+            SOME (lbl, _) =>
+              ( TypeEquality.FromEqTypeHyp lbl orelse_
+                TypeEquality.FromWfHyp lbl orelse_
+                TypeEquality.FromTrueHyp lbl)
```

Comment:
@jonsterling I believe some of these have been implemented somewhere, but I could not find them... so here are my version.